compiler/rustc_lint/src/types.rs

```diff
@@ -832,9 +853,12 @@ impl<'a, 'tcx> ImproperCTypesVisitor<'a, 'tcx> {
         cache: &mut FxHashSet<Ty<'tcx>>,
         field: &ty::FieldDef,
         substs: SubstsRef<'tcx>,
+        allow_unit_type: bool,
     ) -> FfiResult<'tcx> {
         let field_ty = field.ty(self.cx.tcx, substs);
-        if field_ty.has_opaque_types() {
+        if field_ty.is_unit() && allow_unit_type {
```

Comment:
This only checks for unit, shouldn't it be the same check as above? (ZST + no fields + exhaustive)